Add collateral inputs to wallet types, DB, and API.

jonathanknowles opened the pull request:
### Issue Number

ADP-967 / ADP-1069 / ADP-1039

### Overview

This PR adds all the **_boilerplate_** required for collateral inputs to:
- the wallet database;
- the wallet primitive types;
- the wallet API.

This PR makes **_no attempt_** to manipulate collateral inputs in any way, other than to marshall data around from one layer to another. It also makes **_no attempt_** to parse collateral inputs from network blocks. Where necessary, future work has been marked with `TODO` comments. (With a link to ADP-957).

Currently, all API calls returning collateral inputs will return the **_empty list_**.

## API

This PR adds API entries and types for collateral inputs, with the following differences from ordinary inputs:
- while a transaction must have at least one ordinary input, it's possible for a transaction to have **no** collateral inputs.
- while an ordinary input may be associated with one or more non-ada assets, a collateral input **may not** have non-ada assets.

This is reflected in the following differences between types:
```diff
- x-transactionInputs: &transactionInputs
+ x-transactionCollateral: &transactionCollateral
    description: |
-     A list of transaction inputs.
+     A list of transaction inputs used for collateral.
    type: array
    minItems: 0
    items:
      type: object
      required:
        - id
        - index
      properties:
        address: *addressId
        amount: *amount
-       assets: *walletAssets
        id: *transactionId
        index:
          type: integer
          minimum: 0
```

```diff
- x-transactionResolvedInputs: &transactionResolvedInputs
+ x-transactionResolvedCollateral: &transactionResolvedCollateral
-   description: A list of transaction inputs.
+   description: A list of transaction inputs used for collateral.
    type: array
-   minItems: 1
+   minItems: 0
    items:
      type: object
      required:
        - id
        - index
        - address
        - amount
        - derivation_path
      properties:
        address: *addressId
        amount: *amount
-       assets: *walletAssets
        id: *transactionId
        derivation_path: *derivationPath
        index:
          type: integer
          minimum: 0
```

### Database

This PR adds a `TxCollateral` table, with analagous fields to the `TxIn` table.

### Primitive Types

This PR adds fields for collateral inputs to the following primitive types:
  - `UnsignedTx`
  - `TransactionInfo`
  - `Tx`

jonathanknowles committed:
Add collateral input fields to primitive transaction types.

This commit performs the minimum amount of work necessary to add fields
for collateral inputs to the following types:

    - `Tx`
    - `UnsignedTx`
    - `TransactionInfo`

jonathanknowles committed:
Add collateral inputs to the database.

jonathanknowles committed:
Add collateral inputs to `ApiCoinSelection`.

jonathanknowles committed:
Add collateral inputs to `ApiTransaction`.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1920,18 +1924,38 @@ mkApiTransactionFromInfo
         Expired  -> apiTx
   where
-      drop2nd (a,_,c) = (a,c)
+    TransactionInfo
```
Perhaps we could just use `#view` to refer to each field instead of record wild-card syntax.

jonathanknowles replied:
Fixed in b91d00ecde1bbc4bcbeedecf3c0534cf0b0ebf0d.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -2828,6 +2867,16 @@ instance EncodeAddress n => ToJSON (ApiTxInput n) where
+        Object (maybe mempty (fromValue . toJSON) s <> fromValue (toJSON i))
+      where
+        fromValue (Object o) = o
+        fromValue _ = mempty
```
These instances are basically copies of the `ApiTxInput` instances. Perhaps we can simplify them, and use a more generic deriving style.

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -2165,13 +2219,24 @@ selectTxs = fmap concatUnzip . mapM select . chunksOf chunkSize
 
+        resolvedCollateralInputs <- fmap toOutputMap $
+            combineChunked collateralInputs $ \collateralInputsChunk ->
+                traverse readTxOutTokens . fmap entityVal =<< selectList
```
Note that we don't actually expect to read any tokens for collateral inputs.

We should consider simplifying this logic.

sevanspowell commented on specifications/api/swagger.yaml:
```diff
@@ -1212,6 +1230,27 @@ x-transactionResolvedInputs: &transactionResolvedInputs
         minimum: 0
 
+x-transactionResolvedCollateral: &transactionResolvedCollateral
+  description: A list of transaction inputs used for collateral
```
What does "resolved" here mean?

jonathanknowles replied:
> What does "resolved" here mean?

Good question. I'm following the existing terminology for inputs, as much as possible.

For "resolved inputs": in addition to including the `txId` and `txIndex` (which is enough to uniquely identify an input), we also include the following information, which shows what this input "resolves", or "maps" to:
- the `address`
- the `derivationPath`
- the `amount`

jonathanknowles committed:
Use generic field accessors in `mkApiTransactionFromInfo`.

This commit replaces the use of named field puns with generic field
accessors in order to make the function a bit more concise.

In response to:

https://github.com/input-output-hk/cardano-wallet/pull/2817/files#r687606529

jonathanknowles committed:
Use generic field accessors in `mkApiCoinSelection`.

This commit replaces the use of named field puns with generic field
accessors in order to make the function a bit more concise.

In response to:

https://github.com/input-output-hk/cardano-wallet/pull/2817/files#r687606529

sevanspowell commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -2182,7 +2247,10 @@ selectTxs = fmap concatUnzip . mapM select . chunksOf chunkSize
         pure
-            ( inputs `resolveWith` resolvedInputs
+            ( inputs
+                `resolveInputWith`resolvedInputs
```
```suggestion
                `resolveInputWith` resolvedInputs
```

jonathanknowles replied:
Fixed in fd0c08234f222349de62cedcf8a66f841af46c60.

jonathanknowles committed:
Use named record fields in `selectionToUnsignedTx`.

This commit uses named record fields rather than a positional style.

jonathanknowles committed:
Unify terminology for collateral inputs.

For brevity, use "collateral" rather than "collateralInputs" in record
field names.

This is consistent with the API naming.

sevanspowell commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2856,7 +2856,7 @@ mkApiTransaction
     -> IO (ApiTransaction n)
 mkApiTransaction
-    ti txid mfee ins cins outs ws (meta, timestamp) txMeta setTimeReference = do
+    ti txid mfee cins ins outs ws (meta, timestamp) txMeta setTimeReference = do
```
Ooh, this looks like a very likely place to make a mistake, is there anyway we can distinguish between the two types of [(TxIn, Maybe TxOut)] that makes sense in this context?

jonathanknowles replied:
@sevanspowell I agree that it would be easy for a caller to get these arguments muddled up, especially given that the types are identical.

I've created a record type `MkApiTransactionParams`, and altered `mkApiTransaction` to take this record instead of all the positional parameters. Hopefully this should make it easier at call sites to determine which parameter is which.

Commit: c9bc8e687068eb4bd0e556596266d22644ae0833

jonathanknowles committed:
Introduce record type `MkApiTransactionParams`.

This change replaces the use of positional arguments with a record type
in function `mkApiTransaction`, making it clearer at call sites which
argument is which, and hopefully less likely for arguments with the same
type to be supplied in the wrong order.
